File: src/WePrint/Models/BidModel.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Threading.Tasks;
using System.Reflection;

namespace WePrint.Models
{
    public class BidModel : DbModel, IIdempotentDbModel
    {
        public string BidderId { get; set; }
        // We store the Job Idempotency Key here so that if it doesn't match the jobs current idemp key, we know this bid is out-of-date and no longer valid.
        public int JobIdempotencyKey { get; set; }
        public string JobId { get; set; }
        public float Price { get; set; }
        public TimeSpan WorkTime { get; set; }
        public string Notes { get; set; }
        public float LayerHeight { get; set; }
        public float ShellThickness { get; set; }
        public float FillPercentage { get; set; }
        public float SupportDensity { get; set; }
        public string PrinterId { get; set; }
        public MaterialType MaterialType { get; set; }
        public MaterialColor MaterialColor { get; set; }
        public FinishType Finishing { get; set; }
        public int IdempotencyKey { get; set; }
        public bool Accepted { get; set; }

        public void ApplyChanges(BidUpdateModel update)
        {
            ReflectionHelper.CopyPropertiesTo(update, this);
        }
    }

    public class NewBidModel
    {
        public string JobId { get; set; }
        public int JobIdempotencyKey { get; set; }
        public float Price { get; set; }
        public TimeSpan WorkTime { get; set; }
        public string Notes { get; set; }
        public float LayerHeight { get; set; }
        public float ShellThickness { get; set; }
        public float FillPercentage { get; set; }
        public float SupportDensity { get; set; }
        public string PrinterId { get; set; }
        public MaterialType MaterialType { get; set; }
        public MaterialColor MaterialColor { get; set; }
        public FinishType Finishing { get; set; }
    }

    public class BidUpdateModel
    {
        public string JobId { get; set; }
        public string Id { get; set; }
        public float? Price { get; set; }
        public TimeSpan? WorkTime { get; set; }
        public string Notes { get; set; }
        public float? LayerHeight { get; set; }
        public float? ShellThickness { get; set; }
        public float? FillPercentage { get; set; }
        public float? SupportDensity { get; set; }
        public string PrinterId { get; set; }
        public MaterialType? MaterialType { get; set; }
        public MaterialColor? MaterialColor { get; set; }
        public FinishType? Finishing { get; set; }
    }
}

File: src/WePrint/ClientApp/src/models/BidModel.js
import TimeModel from './TimeModel';
import { MaterialType, MaterialColor, FinishType } from './Enums';

export default class BidModel {
  constructor() {
    this.id = undefined;
    this.bidderId = null;
    this.jobId = null;
    this.price = 0;
    this.workTime = new TimeModel();
    this.notes = null;
    this.layerHeight = 0;
    this.shellThickness = 0;
    this.fillPercentage = 0;
    this.supportDensity = 0;
    this.printerId = null;
    this.materialType = MaterialType.ABS;
    this.materialColor = MaterialColor.Red;
    this.finishType = FinishType.None;
    this.accepted = false;
  }

  static IdEquals(a, b) {
    return a.id === b.id;
  }

  static AllPropertiesEqual(a, b) {
    return (
      BidModel.IdEquals(a, b) &&
      a.bidderId === b.bidderId &&
      a.jobId === b.jobId &&
      a.price === b.price &&
      TimeModel.AllPropertiesEqual(a.workTime, b.workTime) &&
      a.notes === b.notes &&
      a.layerHeight === b.layerHeight &&
      a.shellThickness === b.shellThickness &&
      a.fillPercentage === b.fillPercentage &&
      a.supportDensity === b.supportDensity &&
      a.printerId === b.printerId &&
      a.materialType === b.materialType &&
      a.materialColor === b.materialColor &&
      a.finishType === b.finishType &&
      a.accepted === b.accepted
    );
  }
}
